### PR TITLE
tcp_socket: Add secure tcp_socket implementation

### PR DESCRIPTION
**Pull request by mchindri**
Add support for TLS communication over TCP socket using mbedtls

Signed-off-by: Mihail Chindris <mihail.chindris@analog.com>

### CODE REVIEW

File: network/transport/tcp_socket.c [hunk 1]
```diff
@@ -109,7 +267,8 @@ int32_t socket_remove(struct tcp_socket_desc *desc)
 
 	if (!desc)
 		return FAILURE;
-
+	if (desc->secure)
+		stcp_socket_remove(desc->secure);
```
Comment (cristiantpop) [inline network/transport/tcp_socket.c:271]: Add new line after on line body.

File: network/transport/tcp_socket.c [hunk 2]
```diff
@@ -45,23 +45,169 @@
...
+	/* Set default configuration: TLS client socket */
+	ret = mbedtls_ssl_config_defaults(&ldesc->conf,
+					  MBEDTLS_SSL_IS_CLIENT,
+					  MBEDTLS_SSL_TRANSPORT_STREAM,
+					  MBEDTLS_SSL_PRESET_DEFAULT);
+	if (IS_ERR_VALUE(ret))
+		goto exit;
+
+	/* Certificate validation is set as mandatory */
+	mbedtls_ssl_conf_authmode(&ldesc->conf, MBEDTLS_SSL_VERIFY_NONE);
```
Comment (adrimbarean) [inline network/transport/tcp_socket.c:154]: If certificate check is mandatory and this is a client shouldn't the parameter be `MBEDTLS_SSL_VERIFY_REQUIRED`? [Function comments](https://github.com/ARMmbed/mbedtls/blob/5d318cc22b30235d580528c85b7e52e8c565695c/include/mbedtls/ssl.h#L1482) seem to suggest so.

Comment (mchindri) [reply network/transport/tcp_socket.c:154]: It didn't work with verification (maybe I have to fix some configuration).
It is in the next steps from https://github.com/analogdevicesinc/no-OS/pull/586
I will solve this and the rest in other pull request.
This is a first functional version.

**Comment by mchindri** [comment]
v2:
 Remove extra line

File: network/transport/tcp_socket.c (hunk 1 above)
Comment (mchindri) [reply network/transport/tcp_socket.c:270]: done

File: network/transport/tcp_socket.c (hunk 2 above)
Comment (adrimbarean) [reply network/transport/tcp_socket.c:154]: Maybe you need to call `mbedtls_ssl_config_init` after `mbedtls_ssl_config_defaults` to actually set the config.

File: network/transport/tcp_socket.c [hunk 3]
```diff
@@ -45,23 +45,168 @@
...
+
+/* Remove secure descriptor*/
+static void stcp_socket_remove(struct secure_socket_desc *desc)
+{
+	mbedtls_x509_crt_free(&desc->ca_cert);
+	mbedtls_x509_crt_free(&desc->client_cert);
+	mbedtls_ssl_config_free(&desc->conf);
+	if (desc->trng)
+		trng_remove(desc->trng);
+	free(desc);
```
Comment (cristiantpop) [inline network/transport/tcp_socket.c:116]: New line after single line if statement body

File: network/transport/tcp_socket.c (hunk 2 above)
Comment (mchindri) [reply network/transport/tcp_socket.c:154]: init should be called befora any use of this structures. It just set the memory to 0. This is how is done in the library examples also.

File: network/transport/tcp_socket.c (hunk 3 above)
Comment (mchindri) [reply network/transport/tcp_socket.c:116]: done

**Comment by mchindri** [comment]
v3
 - Add new line after single line if body.

**Comment by mchindri** [comment]
v4:
 - Fix astyle
Will merge today if no more comments

**Pull request merged**